src/main/java/dinkplugin/message/DiscordMessageHandler.java

```diff
@@ -98,116 +102,167 @@ public void createMessage(String webhookUrl, boolean sendImage, @NonNull Notific
-                } else {
-                    log.debug("Skipping retry attempts for failed webhook since max retries is not positive");
+
+                    // retry with no change to NotificationBody
+                    retry.accept(mBody, new RuntimeException(String.format("Received unsuccessful http response: %d - %s - %s", response.code(), response.message(), body)));
                 }
+            } catch (Exception e) {
+                retry.accept(mBody, e);
             }
         });
     }
 
+    private NotificationBody<?> enrichBody(NotificationBody<?> mBody, boolean sendImage) {
+        if (mBody.getPlayerName() == null) {
+            mBody = mBody.withPlayerName(Utils.getPlayerName(client));
+        }
+
+        if (mBody.getAccountType() == null) {
+            mBody = mBody.withAccountType(Utils.getAccountType(client));
+        }
+
+        NotificationBody.NotificationBodyBuilder<?> builder = mBody.toBuilder();
+
+        if (config.sendDiscordUser()) {
+            builder.discordUser(DiscordProfile.of(discordService.getCurrentUser()));
+        }
+
+        if (config.sendClanName()) {
+            ClanChannel clan = client.getClanChannel(ClanID.CLAN);
+            if (clan != null) {
+                builder.clanName(clan.getName());
+            }
+        }
+
+        if (config.sendGroupIronClanName()) {
+            ClanChannel gim = client.getClanChannel(ClanID.GROUP_IRONMAN);
+            if (gim != null) {
+                builder.groupIronClanName(gim.getName());
+            }
+        }
+
+        if (config.discordRichEmbeds()) {
+            builder.embeds(computeEmbeds(mBody, sendImage, config));
+        } else {
+            builder.computedDiscordContent(mBody.getText().evaluate(false));
+        }
+
+        return builder.build();
+    }
+
+    private NotificationBody<?> injectThreadName(HttpUrl url, NotificationBody<?> mBody, boolean force) {
+        Collection<String> queryParams = url.queryParameterNames();
+        if (force || (queryParams.contains("forum") && !queryParams.contains("thread_id"))) {
+            String threadName = Template.builder()
+                .template(config.threadNameTemplate())
+                .replacementBoundary("%")
+                .replacement("%TYPE%", Replacements.ofText(mBody.getType().getTitle()))
+                .replacement("%MESSAGE%", mBody.getText())
+                .replacement("%USERNAME%", Replacements.ofText(mBody.getPlayerName()))
+                .build()
+                .evaluate(false);
+            return mBody.withThreadName(Utils.truncate(threadName, NotificationBody.MAX_THREAD_NAME_LENGTH));
+        }
+        return mBody;
+    }
+
+    private MultipartBody createBody(NotificationBody<?> mBody, @Nullable RequestBody image) {
+        MultipartBody.Builder requestBody = new MultipartBody.Builder()
+            .setType(MultipartBody.FORM)
+            .addFormDataPart("payload_json", gson.toJson(mBody));
```

Comment:
This is less performant since we do json serialization separately for each url & retry, but this can be (re)optimized in a follow-up PR (and this work doesn't block the main thread anyways)